sematic/db/queries.py

```diff
@@ -146,6 +194,19 @@ def save_graph(runs: List[Run], artifacts: List[Artifact], edges: List[Edge]):
         session.commit()
 
 
+def _assert_allowed_run_modifications(runs):
+    with db().get_session() as session:
+        for run in runs:
+            existing_run = session.query(Run).filter(Run.id == run.id).one_or_none()
+            if existing_run is not None:
+                if len(run.external_jobs) < len(existing_run.external_jobs):
+                    raise ValueError(
+                        f"Cannot remove existing external jobs from {run.id}. "
+                        f"Existing run had: {existing_run.external_jobs}. New "
+                        f"run had: {run.external_jobs}"
+                    )
```

Comment:
This creates N+1 queries. Can we optimize? Load all runs at once? Also can we only load (id, external_jobs) instead of the entire model?

Also, can we name the function more specifically, e.g. `_assert_no_race_condition` or something (bad example, but you get the point).

Reply:
updated!